agent_build_refactored/tools/run_in_ec2/boto3_tools.py

```diff
@@ -0,0 +1,445 @@
+        )
+
+    def create_boto3_session(self):
+        import boto3
+
+        return boto3.session.Session(
+            aws_access_key_id=self.access_key,
+            aws_secret_access_key=self.secret_key,
+            region_name=self.region,
+        )
+
+
+def create_and_deploy_ec2_instance(
+        boto3_session: boto3.session.Session,
+        ec2_image: EC2DistroImage,
+        name_prefix: str,
+        aws_settings: AWSSettings,
+        root_volume_size: int = None,
+        files_to_upload: Dict = None,
+        deployment_script: pl.Path = None
+):
+    """
+    Create AWS EC2 instance and additional deploy files or scripts.
+    :param boto3_session: Boto3 session.
+    :param ec2_image: AMI image which is used to start instance.
+    :param name_prefix: Additional name prefix for instance.
+    :param aws_settings: All required AWS settings and credentials.
+    :param root_volume_size: Size of root volume on GB
+    :param files_to_upload:
+    :param deployment_script:
+    :return:
+    """
+    add_current_ip_to_prefix_list(
+        boto3_session=boto3_session,
+        prefix_list_id=aws_settings.security_groups_prefix_list_id,
+        ec2_objects_name_prefix=aws_settings.ec2_objects_name_prefix
+    )
+
+    name = f"{INSTANCE_NAME_STRING}-{ec2_image.short_name}-{aws_settings.ec2_objects_name_prefix}-{name_prefix}"
+
+    instance = create_ec2_instance(
+        boto3_session=boto3_session,
+        ec2_image=ec2_image,
+        name=name,
+        aws_settings=aws_settings,
+        root_volume_size=root_volume_size
+    )
+
+    # Try to establish SSH connection.
+    attempts = 10
+    while True:
+        try:
+            ssh = create_ssh_connection(
+                ip=instance.public_ip_address,
+                username=ec2_image.ssh_username,
+                private_key_path=aws_settings.private_key_path,
+            )
+            break
+        except paramiko.ssh_exception.NoValidConnectionsError:
+            logger.info(f"Can not establish SSH connection with {instance.public_ip_address}")
+            if attempts == 0:
+                logger.exception("Giving up. Error: ")
+                raise
+
+            attempts -= 1
+            logger.info("    retry in 10 seconds.")
+            time.sleep(10)
+
+    files_to_upload = files_to_upload or {}
+
+    deployment_command = None
+    if deployment_script:
+        remote_deployment_script_path = pl.Path("/tmp") / deployment_script.name
+        files_to_upload[str(deployment_script)] = str(remote_deployment_script_path)
+        deployment_command = ["bash", str(remote_deployment_script_path)]
+
+    if files_to_upload:
+        ssh_put_files(
+            ssh_connection=ssh,
+            files=files_to_upload
+        )
+        if deployment_command:
+            ssh_run_command(
+                ssh_connection=ssh,
+                command=deployment_command
+            )
+
+        ssh.close()
```

Comment:
The ssh.close() should occur even if there are no files_to_upload

Reply:
oops. Thanks.